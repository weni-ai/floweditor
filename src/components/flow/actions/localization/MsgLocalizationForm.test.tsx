import * as React from 'react';

import {
  fireEvent,
  render,
  fireUnnnicTextAreaChangeText,
  fireUnnnicInputChangeText,
  act,
  fireUnnnicTagDelete,
} from 'test/utils';
import {
  getLocalizationFormProps,
  Spanish,
  createSendMsgAction,
  createSayMsgAction,
} from 'testUtils/assetCreators';
import axios from 'axios';

import SendMsgLocalizationForm from './MsgLocalizationForm';
import userEvent from '@testing-library/user-event';

vi.mock('axios');

describe(SendMsgLocalizationForm.name, () => {
  it('renders send msg', () => {
    const props = getLocalizationFormProps(createSendMsgAction());
    const { baseElement } = render(<SendMsgLocalizationForm {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('handles changes', () => {
    const props = getLocalizationFormProps(createSendMsgAction());
    const { getByTestId, getByText } = render(
      <SendMsgLocalizationForm {...props} />,
    );

    // modify the message
    fireUnnnicTextAreaChangeText(getByTestId('Message'), 'translated message');

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('removes translations', () => {
    const action = createSendMsgAction();
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: { text: ['hola'] },
    });

    // show that we initialized with hola
    const { getByTestId, getByText } = render(
      <SendMsgLocalizationForm {...props} />,
    );
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();

    // clear the translation
    fireUnnnicTextAreaChangeText(getByTestId('Message'), '');

    // save our translations, which should remove the key
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should translate quick replies', async () => {
    const action = createSendMsgAction();
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: { text: ['hola'], quick_replies: ['si', 'no'] },
    });

    const { getByTestId, getByText } = render(
      <SendMsgLocalizationForm {...props} />,
    );

    // modify our tab
    fireEvent.click(getByText('Quick Replies'));

    // remove a quick reply
    await act(async () => {
      fireUnnnicTagDelete(getByTestId('temba_remove_tag_no'));
    });

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should not allow invalid quick replies size', () => {
    const action = createSendMsgAction();
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: {
        text: ['hola'],
        quick_replies: [
          '1',
          '2',
          '3',
          '4',
          '5',
          '6',
          '7',
          '8',
          '9',
          '10',
          '11',
        ],
      },
    });

    const { getByTestId, getByText } = render(
      <SendMsgLocalizationForm {...props} />,
    );

    // try to save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).not.toHaveBeenCalled();

    // modify our tab
    fireEvent.click(getByText('Quick Replies'));

    // remove a quick reply
    fireUnnnicTagDelete(getByTestId('temba_remove_tag_11'));

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should translate template without variables', () => {
    const action = createSendMsgAction({
      templating: {
        template: {
          name: 'template',
          uuid: '9c22b594-fcab-4b29-9bcb-ce4404894a80',
        },
        uuid: 'd54ee634-c573-442b-8528-dea2765389a0',
        variables: [],
      },
    });
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: {
        text: ['hola'],
        templating: {
          template: {
            name: 'template',
            uuid: '9c22b594-fcab-4b29-9bcb-ce4404894a80',
          },
          uuid: 'd54ee634-c573-442b-8528-dea2765389a0',
          variables: [],
        },
      },
    });

    const { getByText } = render(<SendMsgLocalizationForm {...props} />);

    // modify our tab
    fireEvent.click(getByText('WhatsApp'));

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should translate template variables', () => {
    const action = createSendMsgAction({
      templating: {
        template: {
          name: 'template',
          uuid: '9c22b594-fcab-4b29-9bcb-ce4404894a80',
        },
        uuid: 'd54ee634-c573-442b-8528-dea2765389a0',
        variables: ['first', 'second'],
      },
    });
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: {
        text: ['hola'],
        templating: {
          template: {
            name: 'template',
            uuid: '9c22b594-fcab-4b29-9bcb-ce4404894a80',
          },
          uuid: 'd54ee634-c573-442b-8528-dea2765389a0',
          variables: ['uno', 'dos'],
        },
      },
    });

    const { getByTestId, getByText } = render(
      <SendMsgLocalizationForm {...props} />,
    );

    // modify our tab
    fireEvent.click(getByText('WhatsApp'));

    // modify our variables
    fireUnnnicInputChangeText(getByTestId('Variable 1'), 'uno nuevo');
    fireUnnnicInputChangeText(getByTestId('Variable 2'), 'dos nuevo');

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should translate attachments', async () => {
    const action = createSendMsgAction({
      attachments: ['image:https://example.com/image.jpg'],
    });
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: {
        text: ['hola'],
        attachments: ['image:https://example.com/image_spanish.jpg'],
      },
    });

    const { getByTestId, getByText, getAllByText, getAllByTestId } = render(
      <SendMsgLocalizationForm {...props} />,
    );

    // modify our tab
    fireEvent.click(getByText('Attachments'));

    // modify our attachment
    await act(async () => {
      fireUnnnicInputChangeText(
        getAllByTestId('URL')[0],
        'https://example.com/image_spanish_modified.jpg',
      );
    });

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();

    // add a new attachment
    userEvent.click(getAllByText('Image URL')[1]);
    await act(async () => {
      fireUnnnicInputChangeText(
        getAllByTestId('URL')[1],
        'https://example.com/image_spanish_new.jpg',
      );
    });

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();

    // modify our tab
    fireEvent.click(getByText('Attachments'));

    // remove our attachment
    await act(async () => {
      fireEvent.click(getByTestId('remove-attachment-1'));
    });

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should translate audio from say_msg action', async () => {
    const action = createSayMsgAction();
    const props = getLocalizationFormProps(action, Spanish, {
      [action.uuid]: { audio_url: 'https://example.com/audio' },
    });

    const { getByTestId, getByText } = render(
      <SendMsgLocalizationForm {...props} />,
    );

    // @ts-expect-error we are mocking axios post request in __mocks__/axios.ts
    axios.post.mockResolvedValue({
      data: { url: ['https://example.com/audio_new.mp3'] },
    });

    // modify our audio
    await act(async () => {
      fireEvent.change(getByTestId('upload-audio-input'), {
        tager: { files: [new File([''], 'audio.mp3')] },
      });
    });

    // save our translations
    fireEvent.click(getByText('Ok'));
    expect(props.updateLocalizations).toMatchSnapshot();
  });

  it('should close the modal', () => {
    const props = getLocalizationFormProps(createSendMsgAction());
    const { getByText } = render(<SendMsgLocalizationForm {...props} />);

    fireEvent.click(getByText('Cancel'));
    expect(props.onClose).toHaveBeenCalled();
  });
});
