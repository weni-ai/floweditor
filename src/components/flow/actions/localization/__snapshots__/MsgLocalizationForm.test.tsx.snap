// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MsgLocalizationForm > handles changes 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "eng",
      [
        {
          "translations": {
            "attachments": [],
            "text": "translated message",
          },
          "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > removes translations 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "text": "hola",
          },
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > removes translations 2`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "text": "hola",
          },
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
      ],
    ],
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
          },
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > renders send msg 1`] = `
<body>
  <div>
    <div
      class="_dialog_af1b5d"
    >
      <div
        class="_header_af1b5d"
      >
        <div
          class="_title_container_af1b5d"
        >
          <div
            class="_title_af1b5d u font secondary title-sm black"
          >
            Send message
          </div>
          <div
            class="_subtitle_af1b5d"
          />
        </div>
        <div
          data-use-vue-component-wrap=""
          data-v-app=""
          style="all: unset;"
        >
          <div
            class="tab size-sm _tabs_af1b5d"
            data-v-1389f531=""
          >
            <header
              class="tab-header"
              data-v-1389f531=""
            >
              <ul
                class="tab-content"
                data-v-1389f531=""
              >
                
                <li
                  class="tab-head tab-head--active"
                  data-v-1389f531=""
                >
                  
                  <div
                    __use_react_component_wrap=""
                    style="all: unset;"
                  >
                    General
                  </div>
                  
                  <!---->
                </li>
                <li
                  class="tab-head"
                  data-v-1389f531=""
                >
                  
                  <div
                    __use_react_component_wrap=""
                    style="all: unset;"
                  >
                    Attachments
                  </div>
                  
                  <!---->
                </li>
                <li
                  class="tab-head"
                  data-v-1389f531=""
                >
                  
                  <div
                    __use_react_component_wrap=""
                    style="all: unset;"
                  >
                    Quick Replies
                  </div>
                  
                  <!---->
                </li>
                
              </ul>
            </header>
            <main
              class="tab-body"
              data-v-1389f531=""
            >
              <div
                class="tab-panel"
                data-v-1389f531=""
              >
                
                
              </div>
            </main>
          </div>
        </div>
      </div>
      <div
        class="_content_af1b5d"
      >
        <div
          data-spec="translation-container"
        >
          <div
            class="_translate_from_604508"
            data-spec="text-to-translate"
          >
            Hey!
          </div>
        </div>
        <div>
          <div
            class="_completion_wrapper_17662c"
          >
            <div
              data-use-vue-component-wrap=""
              data-v-app=""
              style="all: unset;"
            >
              <div
                class="unnnic-text-area sm normal _textarea_17662c"
                data-testid="Message"
                data-v-82a4b073=""
              >
                <!---->
                <textarea
                  data-v-82a4b073=""
                  placeholder="English Translation"
                  value=""
                />
                <!---->
                <!---->
              </div>
            </div>
            <div
              class="_options_wrapper_08dd12"
              data-testid="temba_completion_options_Message"
              id=""
              style="margin-top: 1px; width: -24px; display: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="_footer_af1b5d"
      >
        <div
          class="_buttons_af1b5d"
        >
          <div
            class="_buttons_wrapper_af1b5d"
          >
            <div
              class="_right_buttons_af1b5d"
            >
              <div
                data-use-vue-component-wrap=""
                data-v-app=""
              >
                <button
                  class="unnnic-button unnnic-button--size-large unnnic-button--tertiary"
                  data-v-4cec5e43=""
                  style="margin-left: 0px; margin-top: 0px; margin-right: 0px;"
                >
                  <!---->
                  <!---->
                  <!---->
                  <span
                    class="unnnic-button__label"
                    data-v-4cec5e43=""
                  >
                    
                    
                     Cancel
                  </span>
                  <!---->
                </button>
              </div>
              <div
                data-use-vue-component-wrap=""
                data-v-app=""
              >
                <button
                  class="unnnic-button unnnic-button--size-large unnnic-button--primary"
                  data-v-4cec5e43=""
                  style="margin-top: 0px; margin-right: 0px;"
                >
                  <!---->
                  <!---->
                  <!---->
                  <span
                    class="unnnic-button__label"
                    data-v-4cec5e43=""
                  >
                    
                    
                     Ok
                  </span>
                  <!---->
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MsgLocalizationForm > should not allow invalid quick replies size 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "quick_replies": [
              "1",
              "2",
              "3",
              "4",
              "5",
              "6",
              "7",
              "8",
              "9",
              "10",
            ],
            "text": "hola",
          },
          "uuid": "dd57500e-a8e8-4a98-8b5b-a33637523a8d",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate attachments 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [
              "image:https://example.com/image_spanish_modified.jpg",
            ],
            "text": "hola",
          },
          "uuid": "d2b146a5-da3b-4766-9d5a-10c91c09dce3",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate attachments 2`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [
              "image:https://example.com/image_spanish_modified.jpg",
            ],
            "text": "hola",
          },
          "uuid": "d2b146a5-da3b-4766-9d5a-10c91c09dce3",
        },
      ],
    ],
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [
              "image:https://example.com/image_spanish_modified.jpg",
              "image:https://example.com/image_spanish_new.jpg",
            ],
            "text": "hola",
          },
          "uuid": "d2b146a5-da3b-4766-9d5a-10c91c09dce3",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate attachments 3`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [
              "image:https://example.com/image_spanish_modified.jpg",
            ],
            "text": "hola",
          },
          "uuid": "d2b146a5-da3b-4766-9d5a-10c91c09dce3",
        },
      ],
    ],
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [
              "image:https://example.com/image_spanish_modified.jpg",
              "image:https://example.com/image_spanish_new.jpg",
            ],
            "text": "hola",
          },
          "uuid": "d2b146a5-da3b-4766-9d5a-10c91c09dce3",
        },
      ],
    ],
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [
              "image:https://example.com/image_spanish_modified.jpg",
            ],
            "text": "hola",
          },
          "uuid": "d2b146a5-da3b-4766-9d5a-10c91c09dce3",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate audio from say_msg action 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "audio_url": [
              "https://example.com/audio_new.mp3",
            ],
          },
          "uuid": "3a45258c-914d-46b9-b018-5ea41669a230",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate quick replies 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "quick_replies": [
              "si",
            ],
            "text": "hola",
          },
          "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate template variables 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "text": "hola",
          },
          "uuid": "57bf1459-14b9-4a73-b579-9d1b56f0b276",
        },
        {
          "translations": {
            "variables": [
              "uno nuevo",
              "dos nuevo",
            ],
          },
          "uuid": "d54ee634-c573-442b-8528-dea2765389a0",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`MsgLocalizationForm > should translate template without variables 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "attachments": [],
            "text": "hola",
          },
          "uuid": "c99dcf07-2ad6-4730-9e99-810a39f5a58d",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
